### ISSUE TITLE
[BUG] TorchRec divides the Row-Wise dgrad passed to FBGEMM_GPU by world_size  

### ISSUE THREAD

**Issue by JacoCheung** [meta-pytorch/torchrec#2251]
# 🐛 Describe the bug
When using DMP to row-wise shard a table on multiple GPUs, the data gradient passed to FBGEMM_GPU are divided by the world size. 

For example, If there are 2 GPUs, the num_embeddings is 8 and **weights are zero init'd**.  The lookup keys and gradient used for backward are init'd as below. 

|               |                  Rank0                   |                  rank1                   |
| :-----------: | :--------------------------------------: | :--------------------------------------: |
|  forward Key  |                 `[0,1]`                  |                 `[2,3]`                  |
| backward grad | `[[0.5,0.5,...,0.5],[0.5,0.5,...,0.5]] ` | `[[0.5,0.5,...,0.5],[0.5,0.5,...,0.5]] ` |

Let‘s assume the optimizer is `SGD `and `lr=1`. After backward, the embedding weights for key `[0,1,2,3]` are updated as all `-0.25`. Which is unexpected.

I used `register_full_backward_pre_hook` to add a hook to the TBE module to double check the input gradient, and it's consistent to what I observed. There are more findings:

1. When the world size is 4, the gradient then is `0.125` -- As what the issue tiltle indicates.
2. The gradient captured by the pre_hook is always divided despite of the Optimizer. (I tried adam)

# software version
- TorchRec v0.7.0
# steps to reproduce 
1. Prepare the script

```
import torch
import os
import torchrec
import torch.distributed as dist
from fbgemm_gpu.split_embedding_configs import EmbOptimType, SparseType
from torchrec.modules.embedding_configs import (
    EmbeddingConfig,
    dtype_to_data_type,
)
from fbgemm_gpu.split_embedding_configs import EmbOptimType


dist.init_process_group("nccl")
rank = dist.get_rank()#int(os.environ["LOCAL_RANK"])
world_size = dist.get_world_size() #int(os.environ["WORLD_SIZE"])
# print(f"rank {rank}, world size {world_size}")

local_rank = int(os.environ["LOCAL_RANK"])
world_size = int(os.environ["WORLD_SIZE"])

torch.cuda.set_device(local_rank)
device = torch.device(f"cuda:{local_rank}")
num_embeddings = 8
embedding_dim = 4
def init_fn(input : torch.Tensor):
  with torch.no_grad():
    input.fill_(0)
ec = torchrec.EmbeddingCollection(
    device=torch.device("meta"),
    tables=[
        torchrec.EmbeddingConfig(
            name="product_table",
            embedding_dim=embedding_dim,
            num_embeddings=num_embeddings,
            feature_names=["product"],
            data_type=dtype_to_data_type(torch.float32),
            init_fn = init_fn,
        )
    ],
)

optimizer_kwargs = {
    "optimizer": EmbOptimType.EXACT_SGD,
    "weight_decay": 0,
    "beta1": 0.7,
    "beta2": 0.8,
    "learning_rate": 1,
    "output_dtype": SparseType.FP32,
}

model = torchrec.distributed.DistributedModelParallel(
    ec,
    device=device,
    sharders=[EmbeddingCollectionSharder(optimizer_kwargs)], # 
)
def bp_hook(module, grad_output):
    print(f"[{local_rank}] module {module} shape {grad_output[0].shape} grad_output[0] is {grad_output[0]}")
    return grad_output
model.module._lookups[0]._emb_modules[0]._emb_module.register_full_backward_pre_hook(bp_hook)
if local_rank == 0:
    print(f"plan is {model.plan}")
torch.set_printoptions(precision=4)

for key in [[0,1,2,3],[0,1,2,3]]: 
    # lookup id = 0
    start = local_rank * len(key) // world_size 
    end = (local_rank + 1) * len(key) // world_size 
    key = key[start:end]
    kjt = torchrec.KeyedJaggedTensor(
        keys=["product"],
        values=torch.tensor(key,dtype=torch.int64).cuda(),
        lengths=torch.ones(len(key), dtype=torch.int64).cuda(),
    )
    res = model(kjt)["product"].values()
    grad = torch.ones_like(res) * 0.5
    # EC backward+optimizer.step
    res.backward(grad)
```
2. Run the script with torchrun
`torchrun --nnodes 1 --nproc_per_node 2 <script.py>`

 ## the output of the above 
 
 1. **mode.plan**  

```
    param     | sharding type | compute kernel | ranks 
------------- | ------------- | -------------- | ------
product_table | row_wise      | fused          | [0, 1]

    param     | shard offsets | shard sizes |   placement  
------------- | ------------- | ----------- | -------------
product_table | [0, 0]        | [4, 4]      | rank:0/cuda:0
product_table | [4, 0]        | [4, 4]      | rank:1/cuda:1
```

2. hook output
```
[1] module SplitTableBatchedEmbeddingBagsCodegen() shape torch.Size([0, 4]) grad_output[0] is 
tensor([], device='cuda:1', size=(0, 4))

[0] module SplitTableBatchedEmbeddingBagsCodegen() shape torch.Size([4, 4]) grad_output[0] is 
tensor([[0.2500, 0.2500, 0.2500, 0.2500],
        [0.2500, 0.2500, 0.2500, 0.2500],
        [0.2500, 0.2500, 0.2500, 0.2500],
        [0.2500, 0.2500, 0.2500, 0.2500]], device='cuda:0')
```

As what is shown,  on rank 1, the gradient values are halved.

**Comment by JacoCheung** [comment]
In general, a pure DDP should average reduce gradient given the loss is mean. 
But for row-wise sharding (as opposed to DDP), why the grad is divided too?

**Comment by JacoCheung** [comment]
Close as fixed.
